File: current_session/python/1746.py
from typing import List
class Solution:
    def maxSumAfterOperation(self, nums: List[int]) -> int:
        # naive approach
        # subarray can be partial

        globalMaxSum = float('-inf')
        localMaxSumNoFlip, localMaxSumFlipped = 0, 0
        a, b= [],[]
        for i in range(len(nums)):
            localMaxSumNoFlip, localMaxSumFlipped = max(localMaxSumNoFlip+nums[i], nums[i]), max(localMaxSumNoFlip+(nums[i]*nums[i]), localMaxSumFlipped+nums[i], (nums[i]*nums[i]))
            a.append(localMaxSumNoFlip)
            b.append(localMaxSumFlipped)
            # print('a',a)
            # print('b',b)
            globalMaxSum = max(localMaxSumFlipped, localMaxSumNoFlip, globalMaxSum)
        return globalMaxSum

print(Solution().maxSumAfterOperation([2,-1,-4,-3]))
print(Solution().maxSumAfterOperation([1,-1,1,1,-1,-1,1]))

File: current_session/c++/1746.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  int maxSumAfterOperation(vector<int>& nums) {
    // use DP
    int localSumWithNoSq = nums[0];
    int localSumWithSq = nums[0]*nums[0];
    int globalSum = max(localSumWithSq, localSumWithNoSq);

    for (int i=1; i<nums.size(); ++i) {
      int x = nums[i]*nums[i];
      localSumWithSq = max({x, localSumWithNoSq+x, localSumWithSq+nums[i]});
      localSumWithNoSq = max(localSumWithNoSq+nums[i], nums[i]);
      globalSum = max({globalSum, localSumWithSq, localSumWithNoSq});
    }
    return globalSum;
  }
};


// clang++ -std=c++17 c++/1746.cpp -o 1746 && ./1746 -v
int main() {
  vector<int> test1 = {2,-1,-4,-3};
  vector<int> test2 = {1,-1,1,1,-1,-1,1};
  cout << Solution().maxSumAfterOperation(test1) << endl;
  cout << Solution().maxSumAfterOperation(test2) << endl;
}